### ISSUE TITLE
Crash when 'tmp' directory is on a different partition to 'cache' directory

### ISSUE THREAD

**Issue by ssssam** [devcurmudgeon/ybd#83]
I edited the `ybd.def` file to put my artifacts and git repos into /home/build instead of the default path, because I had more disk space in the /home partition.

```
artifacts: /home/build/artifacts
base: /src
base-path: [/usr/bin, /bin, /usr/sbin, /sbin]
cache-server: http://git.baserock.org:8080/1.0/sha1s?
caches: /src/cache
ccache_dir: /home/build/ccache
defs-schema: ./schema/definitions-schema.json
deployment: /src/tmp/deployments
gits: /home/build/gits
json-schema: ./schema/json-schema.json
no-ccache: false
no-distcc: true
server: http://192.168.56.102:8000/
tar-url: http://git.baserock.org/taballs
tmp: /src/tmp
```

YBD started OK, but the build failed on the second chunk:

```
2015-06-16 18:53:08 [systems/build-system-x86_64.morph] Using <module 'sandboxlib.linux_user_chroot' from '/usr/lib/python2.7/site-packages/sandboxlib/linux_user_chroot.pyc'> for sandboxing
2015-06-16 18:53:08 [build-system-x86_64] Starting assembly
2015-06-16 18:53:09 [python-cliapp] Starting assembly
2015-06-16 18:53:09 [cliapp] Starting assembly
2015-06-16 18:53:09 [python-coverage-test-runner] Starting assembly
2015-06-16 18:53:09 [python-coveragepy] Starting assembly
2015-06-16 18:53:09 [python-core] Starting assembly
2015-06-16 18:53:09 [mako] Starting assembly
2015-06-16 18:53:09 [python-setuptools] Starting assembly
2015-06-16 18:53:09 [core] Starting assembly
2015-06-16 18:53:09 [sqlite3] Starting assembly
2015-06-16 18:53:09 [build-essential] Starting assembly
2015-06-16 18:53:09 [glibc] Starting assembly
2015-06-16 18:53:09 [stage2-fake-bash] Starting assembly
2015-06-16 18:53:09 [stage2-fake-bash] Start build
2015-06-16 18:53:13 [stage2-fake-bash] Git checkout upstream:bash in /src/tmp/tmpWdP2sD/stage2-fake-bash.build
2015-06-16 18:53:13 [stage2-fake-bash] Upstream version 3590145a bash-4.3-27-g3590145 (bash-4.3 + 27 commits)
2015-06-16 18:53:13 [stage2-fake-bash] Logging build commands to /home/build/artifacts/stage2-fake-bash.52d635bc2a14edd105e2c787579f57de00c10c80be8b171d9e19b75d9288b2a0.build-log
2015-06-16 18:53:13 [stage2-fake-bash] Running install-commands
2015-06-16 18:53:13 [stage2-fake-bash] Running command:
printf '#!/bin/sh\nexec /bin/sh "$@"\n' | install -D /proc/self/fd/0 -m 755 "$DESTDIR$PREFIX/bin/bash"
2015-06-16 18:53:17 [stage2-fake-bash] Running strip-commands
2015-06-16 18:53:17 [stage2-fake-bash] Running command:
find "$DESTDIR" -type f \
'(' -perm -111 -o -name '*.so*' -o -name '*.cmxs' -o -name '*.node' ')' \
-exec sh -ec \
'read -n4 hdr <"$1" # check for elf header
if [ "$hdr" != "$(printf \\x7fELF)" ]; then
exit 0
fi
debugfile="$DESTDIR$PREFIX/lib/debug/$(basename "$1")"
mkdir -p "$(dirname "$debugfile")"
objcopy --only-keep-debug "$1" "$debugfile"
chmod 644 "$debugfile"
strip --remove-section=.comment --remove-section=.note --strip-unneeded "$1"
objcopy --add-gnu-debuglink "$debugfile" "$1"' - {} ';'
2015-06-16 18:53:18 [stage2-fake-bash] Creating cache artifact
2015-06-16 18:53:18 [stage2-fake-bash] Now cached as stage2-fake-bash.52d635bc2a14edd105e2c787579f57de00c10c80be8b171d9e19b75d9288b2a0
2015-06-16 18:53:18 [stage2-fake-bash] Cleaned up /src/tmp/tmpWdP2sD
2015-06-16 18:53:18 [stage2-fake-bash] Elapsed time 00:00:08
2015-06-16 18:53:18 [glibc] Installing stage2-fake-bash.52d635bc2a14edd105e2c787579f57de00c10c80be8b171d9e19b75d9288b2a0
2015-06-16 18:53:18 [glibc] Elapsed time 00:00:08
2015-06-16 18:53:18 [build-essential] Elapsed time 00:00:08
2015-06-16 18:53:18 [sqlite3] Elapsed time 00:00:08
2015-06-16 18:53:18 [core] Elapsed time 00:00:09
2015-06-16 18:53:18 [python-setuptools] Elapsed time 00:00:09
2015-06-16 18:53:18 [mako] Elapsed time 00:00:09
2015-06-16 18:53:18 [python-core] Elapsed time 00:00:09
2015-06-16 18:53:18 [python-coveragepy] Elapsed time 00:00:09
2015-06-16 18:53:18 [python-coverage-test-runner] Elapsed time 00:00:09
2015-06-16 18:53:18 [cliapp] Elapsed time 00:00:09
2015-06-16 18:53:18 [python-cliapp] Elapsed time 00:00:09
2015-06-16 18:53:18 [build-system-x86_64] Elapsed time 00:00:10
2015-06-16 18:53:18 [TOTAL] Elapsed time 00:49:43
2015-06-16 18:53:18 [systems/build-system-x86_64.morph] Finished
Traceback (most recent call last):
File "/home/build/ybd/ybd.py", line 66, in <module>
assemble(defs, app.settings['target'])
File "/home/build/ybd/assembly.py", line 126, in assemble
assemble(defs, subcomponent)
File "/home/build/ybd/assembly.py", line 126, in assemble
assemble(defs, subcomponent)
File "/home/build/ybd/assembly.py", line 118, in assemble
assemble(defs, dependency)
File "/home/build/ybd/assembly.py", line 118, in assemble
assemble(defs, dependency)
File "/home/build/ybd/assembly.py", line 118, in assemble
assemble(defs, dependency)
File "/home/build/ybd/assembly.py", line 126, in assemble
assemble(defs, subcomponent)
File "/home/build/ybd/assembly.py", line 118, in assemble
assemble(defs, dependency)
File "/home/build/ybd/assembly.py", line 118, in assemble
assemble(defs, dependency)
File "/home/build/ybd/assembly.py", line 126, in assemble
assemble(defs, subcomponent)
File "/home/build/ybd/assembly.py", line 118, in assemble
assemble(defs, dependency)
File "/home/build/ybd/assembly.py", line 126, in assemble
assemble(defs, subcomponent)
File "/home/build/ybd/assembly.py", line 119, in assemble
sandbox.install(defs, component, dependency)
File "/home/build/ybd/sandbox.py", line 83, in install
_install(defs, this, component)
File "/home/build/ybd/sandbox.py", line 106, in _install
utils.hardlink_all_files(unpackdir, this['sandbox'])
File "/home/build/ybd/utils.py", line 48, in hardlink_all_files
_process_tree(srcpath, destpath, os.link)
File "/home/build/ybd/utils.py", line 67, in _process_tree
actionfunc)
File "/home/build/ybd/utils.py", line 67, in _process_tree
actionfunc)
File "/home/build/ybd/utils.py", line 67, in _process_tree
actionfunc)
File "/home/build/ybd/utils.py", line 78, in _process_tree
actionfunc(srcpath, destpath)
OSError: [Errno 18] Invalid cross-device link
```

Sorry for lack of formatting in the traceback, it is from the systemd journal which seems to strip indentiation.

It would be nice to have either a successful build, or an error raised as soon as YBD starts in this situation. As it was, this crash occurred a long time after the build started (as it cloned every Git repo before starting building), by which time I'd gone home.

**Comment by devcurmudgeon** [comment]
> It would be nice to have either a successful build, or an error raised as soon as YBD starts in this situation.

agreed. i'd prefer to fix it than error. utils.py is code from morph, irrc. i'll take a look.

**Comment by devcurmudgeon** [comment]
i think this is fixed by @jjardon's recent patch. please re-open if not.
